include/dice/template-library/ranges.hpp

```diff
@@ -2,40 +2,466 @@
+				return self(std::forward<R>(range));
+			}
+		};
+	}// namespace ranges_algo_detail
+
+	/**
+	 * Creates a lazy, pipeable view of a range that excludes elements matching a given value. This adaptor does not modify the source range.
+	 *
+	 * @param value The value to filter out of the source range.
+	 * @param pred Optional binary predicate for comparison. Defaults to `std::ranges::equal_to`.
+	 * @return A range adaptor closure.
+	 */
+	template<typename T, typename Pred = std::ranges::equal_to>
+	constexpr auto remove_element(T &&value, Pred &&pred = Pred{}) {
+		return ranges_algo_detail::remove_element_fn<std::decay_t<T>, std::decay_t<Pred>>(
+				std::forward<T>(value), std::forward<Pred>(pred));
+	}
+}// namespace dice::template_library
+
+// all_equal adaptor
+namespace dice::template_library {
+	namespace ranges_algo_detail {
+		template<typename Pred>
+		struct all_equal_fn {
+			Pred pred_;
+
+			template<typename P>
+			constexpr explicit all_equal_fn(P &&pred)
+				: pred_(std::forward<P>(pred)) {}
+
+			template<std::ranges::input_range R>
+				requires std::copyable<std::ranges::range_value_t<R>>
+			constexpr bool operator()(R &&range) const {
+				if (dice::template_library::empty(range)) {
+					return true;
+				}
+
+				auto it = std::ranges::begin(range);
+				const auto first_element = *(it++);
+
+				// Check if all subsequent elements in the range match the first one.
+				return std::ranges::subrange(it, std::ranges::end(range)) |
+					   dice::template_library::all_of([this, first_element](const auto &current_element) {
+						   return std::invoke(pred_, current_element, first_element);
+					   });
+			}
+
+			template<std::ranges::input_range R>
+				requires std::copyable<std::ranges::range_value_t<R>>
+			friend constexpr auto operator|(R &&range, const all_equal_fn &self)
+					-> decltype(self(std::forward<R>(range))) {
+				return self(std::forward<R>(range));
+			}
+		};
+	}// namespace ranges_algo_detail
+
+	/**
+     * @brief Checks if all elements in a range are equal to each other.
+     *
+     * This is a pipeable, terminal algorithm that returns a boolean. Elements must be copyable for this to work.
+     *
+     * @param pred Optional binary predicate for comparison. Defaults to `std::ranges::equal_to`.
+     * @return true if all elements in the range are equal, otherwise false.
+     */
+	template<typename Pred = std::ranges::equal_to>
+	constexpr auto all_equal(Pred &&pred = Pred{}) {
+		return ranges_algo_detail::all_equal_fn<std::decay_t<Pred>>(
+				std::forward<Pred>(pred));
+	}
+}// namespace dice::template_library
+
+// unique adaptor
+namespace dice::template_library {
+	namespace ranges_algo_detail {
+
+		template<std::ranges::input_range V>
+			requires std::ranges::view<V>
+		struct unique_view : std::ranges::view_interface<unique_view<V>> {
+		private:
+			struct iterator;
+
+			using ValueType = std::decay_t<std::ranges::range_value_t<V>>;
```

Comment:
`decay_t` is not the correct typetrait for this (it is almost never the correct type trait for anything). You want `std::remove_cvref_t`